source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2696 @@
+
+  - ``id``: Required string. Unique name for this entity. The YAML file SHOULD
+    define a `node anchor`_ for this field (e.g.
+    ``id: &collection0 collection0``).
+
+  - ``database``: Required string. Database entity from which this collection
+    will be created. The YAML file SHOULD use an `alias node`_ for a database
+    entity's ``id`` field (e.g. ``database: *database0``).
+
+  - ``collectionName``: Required string. Collection name. The YAML file SHOULD
+    define a `node anchor`_ for this field (e.g.
+    ``collectionName: &collection0Name foo``).
+
+  - ``collectionOptions``: Optional `collectionOrDatabaseOptions`_ object.
+
+.. _entity_session:
+
+- ``session``: Optional object. Defines an explicit ClientSession object.
+
+  The structure of this object is as follows:
+
+  - ``id``: Required string. Unique name for this entity. The YAML file SHOULD
+    define a `node anchor`_ for this field (e.g. ``id: &session0 session0``).
+
+  - ``client``: Required string. Client entity from which this session will be
+    created. The YAML file SHOULD use an `alias node`_ for a client entity's
+    ``id`` field (e.g. ``client: *client0``).
+
+  - ``sessionOptions``: Optional object. Map of parameters to pass to
+    `MongoClient.startSession <../source/sessions/driver-sessions.rst#startsession>`__
+    when creating the session. Supported options are defined in the following
+    specifications:
+
+    - `Causal Consistency <../causal-consistency/causal-consistency.rst#sessionoptions-changes>`__
+    - `Transactions <../transactions/transactions.rst#sessionoptions-changes>`__
+
+    When specifying TransactionOptions for ``defaultTransactionOptions``, the
+    transaction options MUST remain nested under ``defaultTransactionOptions``
+    and MUST NOT be flattened into ``sessionOptions``.
+
+.. _entity_bucket:
+
+- ``bucket``: Optional object. Defines a Bucket object, as defined in the
+  `GridFS <../gridfs/gridfs-spec.rst>`__ spec.
+
+  The structure of this object is as follows:
+
+  - ``id``: Required string. Unique name for this entity. The YAML file SHOULD
+    define a `node anchor`_ for this field (e.g. ``id: &bucket0 bucket0``).
+
+  - ``database``: Required string. Database entity from which this bucket will
+    be created. The YAML file SHOULD use an `alias node`_ for a database
+    entity's ``id`` field (e.g. ``database: *database0``).
+
+  - ``bucketOptions``: Optional object. Additional options used to construct
+    the bucket object. Supported options are defined in the
+    `GridFS <../source/gridfs/gridfs-spec.rst#configurable-gridfsbucket-class>`__
+    specification. The ``readConcern``, ``readPreference``, and ``writeConcern``
+    options use the same structure as defined in `Common Options`_.
+
+.. _entity_stream:
+
+- ``stream``: Optional object. Defines a stream, as defined in the
+  `GridFS <../gridfs/gridfs-spec.rst>`__ spec. Test runners MUST ensure that
+  stream is both readable *and* writable.
```

Comment:
Ruby driver has separate read and write stream classes, therefore I don't see how we can fulfill this requirement given our existing implementation. What is the reason for requiring readable and writable streams being implemented by the same class?

Reply:
The [GridFS spec](https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst) only refers to an "IO Stream" and doesn't differentiate between readable and writable streams. Taking the spec at face value, I would expect the API should allow a driver to download a GridFS file to a stream and then pass that same stream to another API method to upload a separate file to GridFS.

If necessary, we could split this into separate readable and writable stream types, but I'd be curious to know why you can't support a bidirectional stream for your GridFS API.

Reply:
We can investigate providing a single stream class, my remark was only that currently we don't do that. I am not familiar enough with our gridfs code to know why it is the way it is off the top of my head.

It sounds like the proposed language in this PR is consistent with the gridfs spec as currently written, therefore I'm OK with resolving this thread and we'll deal with this situation in Ruby when the time comes.